numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -4008,6 +4008,188 @@ static int
  *****************************************************************************
  */
 
+#ifdef NPY_HAVE_SSE2_INTRINSICS
+
+#include <emmintrin.h>
+
+#ifdef __SSE3__
+#include <pmmintrin.h>
+#define hadd _mm_hadd_pd
+#else
+static __m128d hadd(__m128d ab, __m128d cd)
+{
+    __m128d ac = _mm_unpacklo_pd(ab, cd);
+    __m128d bd = _mm_unpackhi_pd(ab, cd);
+    return _mm_add_pd(ac, bd);
+}
+#endif
+
+static int
+vect_cor3(double * d, double *k, double *o, npy_intp ostride, npy_uintp nd)
+{
+    npy_intp i;
+    /* working on two results per iteration so load kernel twice */
+    __m128d vk1 = _mm_loadu_pd(k);
+    __m128d vk2 = _mm_set_pd(k[0], k[2]);
+    __m128d vk3 = _mm_loadu_pd(&k[1]);
+
+    __m128d vd3 = _mm_loadu_pd(&d[0]);
+    for (i = 0; i < nd - (nd % 2); i+=2) {
+        /* load 6 elements with the last 3 overlapping by 2 with first 3 */
+        __m128d vd1, vd2, md1, md2, md3, r;
+        vd1 = vd3; /* _mm_loadu_pd(&d[i]) */
+        vd3 = _mm_loadu_pd(&d[i + 2]);
+        /*
+         * third entry of the first set is the first entry of second set
+         * first entry of the second set is the second entry of first set
+         */
+        vd2 = _mm_shuffle_pd(vd3, vd1, _MM_SHUFFLE2 (1,0));
+
+        /* now the data is 3 vectors aligned with the double kernel vectors */
+        md1 = _mm_mul_pd(vk1, vd1);
+        md2 = _mm_mul_pd(vk2, vd2);
+        md3 = _mm_mul_pd(vk3, vd3);
+        /* sum down to one vector and store */
+        r = _mm_add_pd(md2, hadd(md1, md3));
+        _mm_storeu_pd(&o[i * ostride], r);
+    }
+    if (nd % 2 == 1) {
+        i = nd - 1;
+        o[i * ostride] = d[i] * k[0] + d[i + 1] * k[1] + d[i + 2] * k[2];
+    }
+    return 1;
+}
+
+static int
+vect_cor5(double * d, double *k, double *o, npy_intp ostride, npy_uintp nd)
+{
+    npy_intp i;
+    __m128d vk1 = _mm_loadu_pd(k);
+    __m128d vk2 = _mm_loadu_pd(&k[2]);
+    __m128d vk3 = _mm_set_pd(k[0], k[4]);
+    __m128d vk4 = _mm_loadu_pd(&k[1]);
+    __m128d vk5 = _mm_loadu_pd(&k[3]);
+
+    __m128d vd4 = _mm_loadu_pd(&d[0]);
+    __m128d vd5 = _mm_loadu_pd(&d[2]);
+    for (i = 0; i < nd - (nd % 2); i+=2) {
+        __m128d vd1, vd2, vd3, md1, md2, md3, md4, md5, r;
+        vd1 = vd4; /* _mm_loadu_pd(&d[i]); */
+        vd2 = vd5; /* _mm_loadu_pd(&d[i + 2]); */
+        vd4 = vd2; /* _mm_loadu_pd(&d[i + 2]); */
+        vd5 = _mm_loadu_pd(&d[i + 4]);
```

Comment:
interestingly on an AMD phenom X2 3 loads is around 10% faster than one load with reusing the 2 registers of the last, not sure why its not spilling to stack.